spot_rl_experiments/spot_rl/envs/open_close_drawer_env.py

```diff
@@ -108,11 +129,102 @@ def compute_distance_to_handle(self):
+        """BD API to open the drawer"""
         raise NotImplementedError
 
+    def bd_open_cabinet_api(self):
+        """BD API to open the cabinet"""
+        command = self.spot.construct_cabinet_task(
+            0.25, force_limit=40, target_angle=1.74, position_control=True
+        )
+        task_duration = 10000000
+        command.full_body_command.constrained_manipulation_request.end_time.CopyFrom(
+            self.spot.robot.time_sync.robot_timestamp_from_local_secs(
+                time.time() + task_duration
+            )
+        )
+        self.spot.command_client.robot_command_async(command)
+        time.sleep(10)
+
+    def open_drawer(self):
+        """Heuristics to open the drawer"""
+        # Get the transformation
+        vision_T_base = self.spot.get_magnum_Matrix4_spot_a_T_b("vision", "body")
+        ee_rotation = self.spot.get_ee_quaternion_in_body_frame()
+
+        # Get the transformation of the gripper
+        vision_T_hand = self.spot.get_magnum_Matrix4_spot_a_T_b("vision", "hand")
+        # Get the location that we want to move to for retracting/moving forward the arm. Pull/push the drawer by 20 cm
+        pull_push_distance = -0.2 if self._mode == "open" else 0.25
+        move_target = vision_T_hand.transform_point(
+            mn.Vector3([pull_push_distance, 0, 0])
+        )
+        # Get the move_target in base frame
+        move_target = vision_T_base.inverted().transform_point(move_target)
+
+        # Retract the arm based on the current gripper location
+        self.spot.move_gripper_to_point(
+            move_target, [ee_rotation.w, ee_rotation.x, ee_rotation.y, ee_rotation.z]
+        )
+
+    def open_cabinet(self):
+        """Heuristics to open the cabinet"""
+        # Get the location of the rotataional axis
+        base_T_hand = self.spot.get_magnum_Matrix4_spot_a_T_b("body", "hand")
+
+        # Assuming that the cabinet door's size is panel_size
+        # Assuming that the door axis is on the left side of the hand
+        panel_size = 0.55
+        if self._cab_door == "right":
+            panel_size = -panel_size
+
+        base_T_hand.translation = base_T_hand.transform_point(
+            mn.Vector3(0.0, 0.0, -panel_size)
+        )
+        target_degree = 70
+        interval = 10
+        # Loop over to create a circular motion for the gripper
+        for cur_ang_in_deg in range(10, target_degree + 10, interval):
+            if cur_ang_in_deg < target_degree:
+                # Keep closing the gripper to grasp the handle tightly
+                self.spot.close_gripper()
+            else:
+                # In the final stage, we open the gripper to let gripper be away from the handle
+                self.spot.open_gripper()
+            # Angle in degree
+            cur_ang = np.deg2rad(cur_ang_in_deg)
+
+            # For right side of the hand
+            if self._cab_door == "right":
+                cur_ang = -cur_ang
```

Comment:
Can we make this for loop through IK interpolations ? Basically using Boston Dynamics API ? like execute trajectory or point A to point B with 10degrees of rotation motion.

Reply:
Hi Tushar! Thank you so much for the information. After checking the document from Boston Dynamics (https://dev.bostondynamics.com/python/examples/inverse_kinematics/readme), I found that this IK service still needs to manually specify target gripper points by doing for-loop on each target. This means that we have to use for-loop for both cases. In addition. I ran this example on real robots. I think the difference between spot.move_gripper() and this IK service is that this IK service also adjusts the body orientation, so that the gripper is easier to reach the target. However, in our open/close cases, I do not think we need this.

I hope this clarifies what we discussed!